### ISSUE TITLE
Long delay (30+ seconds) and high CPU utilitatation when AeroSpace resets workspaces etc. after wake from sleep or add/remove display

### ISSUE THREAD

**Issue by jack4git** [nikitabobko/AeroSpace#604]
[(This]() is not a v15.x regression -- same problem existed in v14.x)

I have a MacBook Pro M1 Max with 32GB. I run tons of apps, the heaviest being Firefox, with perhaps 15-20 windows and 120+ tabs. Memory pressure can increase to 50-65%+ but performance is fine normally. I operate with 3 monitors. Also note, because I find in non-feasible to not see the menu bar of an application on that application's monitor, I do have Displays have separate Spaces enabled.

When I've not rebooted for a few days, AeroSpace tends to get laggy (though this usually corresponds to high memory pressure also so ...). But performance is noticeably slow in specific situations:

1. Plugging in an additional display.
2. Unplugging a display.
3. Especially, waking up the MacBook after it has been in sleep mode, which is basically a superset of adding/removing a display -- AeroSpace seems to be resetting everything. This process can take 30-60 seconds when the system has been running for several days.

What I see during the wake up process:

- Windows / workspaces switching from one display to another, partially displaying, etc.
- AeroSpace menu not displaying at all on a monitor; or displaying a different display-workspace config on each display, with some displays not showing one or two displays but not all three
- Long delays when nothing visible happens
- CPU utilization hitting near 100% (pretty heavy on an M1 Max)

Eventually (30-60 seconds), the correct workspaces and AeroSpace menu state appears on all 3 displays. Note that performance is much better if my system was just rebooted and memory pressure is low.

To be clear, I realize AeroSpace isn't the only code executing when all of the above happens. That said, AeroSpace seems to be at least triggering all the activity which causes things to bog down.

[aerospace.toml.txt](https://github.com/user-attachments/files/17450265/aerospace.toml.txt)

## Checklist

- [x] I've searched for duplicates. My search queries were (please don't forget to search in closed issues too):
  - performance
  - wake
  - sleep
- [x] I've read the guide https://nikitabobko.github.io/AeroSpace/guide
- [x] I've searched in the documentation (https://nikitabobko.github.io/AeroSpace/guide & https://nikitabobko.github.io/AeroSpace/commands). My search queries were:
  - performance
  - wake
  - sleep

```
aerospace CLI client version: 0.15.2-Beta b6cf82771f245ab7349a93baf8709e171537ff58
AeroSpace.app server version: 0.15.2-Beta b6cf82771f245ab7349a93baf8709e171537ff58
```

**Comment by raisjn** [comment]
I also experience hangs and tried to make a few optimizations. In particular, I used instruments to find CPU heavy work and eliminate it, but I think it goes beyond CPU: I can see "hangs" that happen when there is little to no CPU use within Aerospace but when the system has heavy CPU usage. I don't have strong experience with optimizing Mac, so I haven't figured out the cause of hangs (priority inversions? waiting on some blocking calls from APIs? needing threads to differentiate user input from other work?)

What I did:

* used renice to push aerospace to highest priority, this saved a bit of time but it didn't prevent Aerospace from getting starved
* optimized MenuBar (it's called many times when switching workspaces)
* Cache the monitors and invalidate the cache when gcMonitors() is called. sortedMonitors is called frequently and this helps. 

I've been using aerospace for a couple days and the hanging during CPU / RAM contention is very difficult to deal with - it's not reasonable to be unable to switch workspaces for 10+ seconds while CPU work is happening. Typically it would be when something is chewing many CPUs (like a linter, compilation process, etc). One frequent cause was when I would do a release build of Aerospace itself, but I get even larger pauses when I'm moving around in a monorepo and all the linters fire off at once. I have an m3, so it's not under powered by any means. 

I'm not sure this is the same issue as what the reporter above has (RAM contention), but the symptoms are similar (hanging and delays of many seconds)

**Comment by raisjn** [comment]
```

**Comment by raisjn** [comment]
ah, I see https://github.com/nikitabobko/AeroSpace/issues/131 - I will pay attention to this issue

**Comment by jack4git** [comment]
@raisjn thanks, I missed #131 and that does indeed seem to be the same issue. @nikitabobko

**Comment by askurihin** [comment]
Maybe changing the priority of the AeroSpace process with `sudo renice -20 $aerospace_pid` could help.

**Comment by loeffel-io** [comment]
same here

**Comment by d4mr** [comment]
here too, every single keybind takes 5-10 seconds to trigger.

for moving a window to a new workspace:
- 5 seconds to move the window to the new workspace
- 5 more seconds for the workspace indicator on the menu bar to update

even for shifting focus with alt+h/l:
- takes 5 seconds to move focus

**Comment by ipstone** [comment]
similarly here, I accidentally turned off one monitor, but turning it on back again, super slow keybind response. Quiting aerospace and reopen it doesn't seem to help.

**Comment by d4mr** [comment]
Have some more information here:
I restarted my macbook, and also updated to macos 15.2. Everything works instantly now, not sure what did it.

**Comment by parkers0405** [comment]
Okay  I think I found the issue, I am curious to see if I am correct or not. Check if  icloud is stuck on uploading or downloading a small file. If so sign out of iCloud, sign backing ```pkill bird```, or file indexing, and restart your laptop.

Thanks!

**Comment by artur-marchetto** [comment]
I experience this too

**Comment by tadsn3w** [comment]
I had a similar experience here. I was using Aerospace, which became sluggish after leaving my Mac Mini to sleep at night. The next day, it was completely slow. Even quitting it didn’t help; the only solution was a reboot. Then I upgraded to 15.2, and suddenly it worked. I upgraded to 15.3, and it continued to work fine without any sluggishness. Now, I upgraded to 15.4, and the sluggishness returned. So, now if I leave my Mac on with Aerospace running overnight, the next day is extremely slow.

**Comment by nikitabobko** [comment]
Duplicate of #131 and should be fixed in 0.18.x releases

**Comment by snwfdhmp** [comment]
Exact same behavior here :

```
$ aerospace -v
aerospace CLI client version: 0.18.5-Beta 4213dfd9d958dbfe3f801e07d7d5af53303baa75
AeroSpace.app server version: 0.18.5-Beta 4213dfd9d958dbfe3f801e07d7d5af53303baa75
```

```
$ neofetch
                    'c.          ***
                 ,xNMM.          -------------------
               .OMMMMo           OS: macOS 15.3.1 24D70 arm64
               OMMM0,            Host: Mac14,9
     .;loddo:' loolloddol;.      Kernel: 24.3.0
   cKMMMMMMMMMMNWMMMMMMMMMM0:    Uptime: 13 hours, 11 mins
 .KMMMMMMMMMMMMMMMMMMMMMMMWd.    Packages: 228 (brew)
 XMMMMMMMMMMMMMMMMMMMMMMMX.      Shell: zsh 5.9
;MMMMMMMMMMMMMMMMMMMMMMMM:       Resolution: 1512x982
:MMMMMMMMMMMMMMMMMMMMMMMM:       DE: Aqua
.MMMMMMMMMMMMMMMMMMMMMMMMX.      WM: Quartz Compositor
 kMMMMMMMMMMMMMMMMMMMMMMMMWd.    WM Theme: Blue (Light)
 .XMMMMMMMMMMMMMMMMMMMMMMMMMMk   Terminal: iTerm2
  .XMMMMMMMMMMMMMMMMMMMMMMMMK.   Terminal Font: RobotoMonoForPowerline-Light 14
    kMMMMMMMMMMMMMMMMMMMMMMd     CPU: Apple M2 Pro
     ;KMMMMMMMWXXWMMMMMMMk.      GPU: Apple M2 Pro
       .cooc,.    .,coo:.        Memory: 2929MiB / 16384MiB
```
